fblock.go

```diff
@@ -0,0 +1,455 @@
+	}
+
+	expansionSize := varintf.Encode(uint64(fb.Header.ExpansionSize))
+	data := make([]byte, FBlockMinHeaderLen+len(expansionSize)+len(fb.Header.ExpansionBytes))
+	i := 0
+	i += copy(data[i:], factoidBlockChainID[:])
+	i += copy(data[i:], fb.Header.BodyMR[:])
+	i += copy(data[i:], fb.Header.PrevKeyMR[:])
+	i += copy(data[i:], fb.Header.PrevLedgerKeyMR[:])
+
+	binary.BigEndian.PutUint64(data[i:], fb.Header.ExchangeRate)
+	i += 8
+	binary.BigEndian.PutUint32(data[i:], fb.Header.Height)
+	i += 4
+	i += copy(data[i:], expansionSize)
+	// Currently all expansion bytes are stored in the ExpansionBytes. So just write them out
+	i += copy(data[i:], fb.Header.ExpansionBytes)
+	binary.BigEndian.PutUint32(data[i:], fb.Header.TransactionCount)
+	i += 4
+	binary.BigEndian.PutUint32(data[i:], fb.Header.BodySize)
+	i += 4
+
+	return data, nil
+}
+
+// ComputeLedgerKeyMR computes the keymr of the factoid block including transaction
+// signatures.
+func (fb FBlock) ComputeKeyMR() (Bytes32, error) {
+	return fb.computeKeyMR(false)
+}
+
+// ComputeLedgerKeyMR computes the keymr of the factoid block excluding transaction
+// signatures.
+func (fb FBlock) ComputeLedgerKeyMR() (Bytes32, error) {
+	return fb.computeKeyMR(true)
+}
+
+func (fb FBlock) computeKeyMR(ledger bool) (Bytes32, error) {
+	if !fb.IsPopulated() {
+		return Bytes32{}, fmt.Errorf("not populated")
+	}
+
+	leaves := make([][]byte, 2)
+	header, err := fb.MarshalBinaryHeader()
+	if err != nil {
+		return Bytes32{}, err
+	}
+
+	var body [32]byte
+	if ledger {
+		body, err = fb.ComputeLedgerBodyMR()
+	} else {
+		body, err = fb.ComputeBodyMR()
+	}
+	if err != nil {
+		return Bytes32{}, err
+	}
+
+	headerHash := sha256.Sum256(header)
+	if ledger { // Merkle leaves are flipped for the ledger keymr
+		leaves[0] = body[:]
+		leaves[1] = headerHash[:]
+	} else {
+		leaves[0] = headerHash[:]
+		leaves[1] = body[:]
+	}
+
+	tree := merkle.NewTreeWithOpts(merkle.TreeOptions{DoubleOddNodes: true, DisableHashLeaves: true})
+	if err := tree.Generate(leaves, sha256.New()); err != nil {
+		return Bytes32{}, err
+	}
+	root := tree.Root()
+	var keyMR Bytes32
+	copy(keyMR[:], root.Hash)
+	return keyMR, nil
+}
+
+// ComputeBodyMR computes the merkle root of the transactions in the body including
+// their signatures
+func (fb FBlock) ComputeBodyMR() (Bytes32, error) {
+	return fb.computeBodyMR(false)
+}
+
+// ComputeBodyMR computes the merkle root of the transactions in the body excluding
+// their signatures
+func (fb FBlock) ComputeLedgerBodyMR() (Bytes32, error) {
+	return fb.computeBodyMR(true)
+}
+
+// computeBodyMR will calculate the merkle root of all the transactions in the body.
+// If `ledger` is true, signature blocks of the transactions are excluded.
+func (fb FBlock) computeBodyMR(ledger bool) (Bytes32, error) {
+	if !fb.IsPopulated() {
+		return Bytes32{}, fmt.Errorf("not populated")
+	}
+
+	// Transactions + minute markers are included
+	leaves := make([][]byte, len(fb.Transactions)+len(fb.endOfPeriod))
```

Comment:
In my most recent changes, I validate the bodyMR in the unmarshal func. When doing this I avoid additional allocations by building the body elements by takings slices of the incoming data as I parse it. When done I pass that to ComputeEBlockBodyMR or ComputeDBlockBodyMR, which are slightly different. You may be able to reuse those.